File: android/InverterMonitor/app/src/main/java/com/bilalahmad/invertermonitor/ui/screens/reports/RawReadingDetailSheet.kt
package com.bilalahmad.invertermonitor.ui.screens.reports

import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Timeline
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.bilalahmad.invertermonitor.data.models.RawReading
import com.bilalahmad.invertermonitor.ui.components.InfoRow
import com.bilalahmad.invertermonitor.ui.components.SheetCard
import com.bilalahmad.invertermonitor.ui.components.SheetScaffold
import com.bilalahmad.invertermonitor.ui.components.SheetSectionHeader
import com.bilalahmad.invertermonitor.ui.components.StatTile
import com.bilalahmad.invertermonitor.ui.theme.Palette
import kotlin.math.roundToInt

@Composable
fun RawReadingDetailSheet(reading: RawReading, onDismiss: () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = Palette.BackgroundTop,
    ) {
        SheetScaffold(
            icon = Icons.Filled.Timeline,
            iconTint = Palette.InverterAmber,
            title = reading.timestampFormatted,
            subtitle = "Inverter reading",
            onDismiss = onDismiss,
        ) {
            // Power section — 2x2 grid, tinted values to match the live charts.
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                SheetSectionHeader("Power", accent = Palette.InverterAmber)
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    StatTile("Solar", "${reading.solarPower.roundToInt()}", "W", Palette.Solar, Modifier.weight(1f))
                    StatTile("Grid", "${reading.gridPower.roundToInt()}", "W", Palette.Grid, Modifier.weight(1f))
                }
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    StatTile("Load", "${reading.loadPower.roundToInt()}", "W", Palette.Load, Modifier.weight(1f))
                    StatTile("Battery", "${reading.batteryPower.roundToInt()}", "W", Palette.Battery, Modifier.weight(1f))
                }
            }

            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                SheetSectionHeader("Battery", accent = Palette.Battery)
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    StatTile(
                        "State of Charge",
                        "${reading.batteryPercentage.roundToInt()}", "%",
                        batteryColor(reading.batteryPercentage),
                        Modifier.weight(1f),
                    )
                    StatTile(
                        "Direction",
                        directionLabel(reading.batteryPower),
                        accent = Color.White,
                        modifier = Modifier.weight(1f),
                    )
                }
            }

            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                SheetSectionHeader("Environment", accent = Palette.Grid)
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    StatTile("Grid Voltage", "${reading.gridVoltage.roundToInt()}", "V", Color.White, Modifier.weight(1f))
                    StatTile("Inverter Temp", "${reading.temperature.roundToInt()}", "°C", Color.White, Modifier.weight(1f))
                }
            }

            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                SheetSectionHeader("Raw", accent = Palette.SubtleText)
                SheetCard {
                    InfoRow("Timestamp", reading.timestampFormatted)
                    InfoRow("Epoch seconds", "${reading.timestamp.toLong()}")
                    InfoRow("Cycle duration", String.format(java.util.Locale.US, "%.1f ms", reading.durationMs))
                }
            }
        }
    }
}

private fun directionLabel(power: Double): String = when {
    power > 5 -> "Charging"
    power < -5 -> "Discharging"
    else -> "Idle"
}

private fun batteryColor(pct: Double): Color = when {
    pct >= 50 -> Palette.Battery
    pct >= 20 -> Color(0xFFFFA726)
    else -> Color(0xFFFF6B6B)
}

File: ios/InverterMonitor/InverterMonitor/Views/Reports/RawReadingDetailSheet.swift
import SwiftUI

/// Bottom sheet that appears when a row in the Raw readings table is tapped.
/// Shows every field from the reading, not just the four shown in the condensed row.
struct RawReadingDetailSheet: View {
    let reading: RawReading
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 20) {
                    header

                    section("Power (W)") {
                        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 10) {
                            StatTile(label: "Solar", value: "\(Int(reading.solarPower.rounded()))",
                                     unit: "W", accent: Palette.solar)
                            StatTile(label: "Grid", value: "\(Int(reading.gridPower.rounded()))",
                                     unit: "W", accent: Palette.grid)
                            StatTile(label: "Load", value: "\(Int(reading.loadPower.rounded()))",
                                     unit: "W", accent: Palette.load)
                            StatTile(label: "Battery", value: "\(Int(reading.batteryPower.rounded()))",
                                     unit: "W", accent: Palette.battery)
                        }
                    }

                    section("Battery") {
                        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 10) {
                            StatTile(label: "State of Charge",
                                     value: "\(Int(reading.batteryPercentage.rounded()))",
                                     unit: "%",
                                     accent: socColor(reading.batteryPercentage))
                            StatTile(label: "Direction",
                                     value: directionLabel,
                                     accent: .white)
                        }
                    }

                    section("Grid") {
                        LazyVGrid(columns: [GridItem(.flexible()), GridItem(.flexible())], spacing: 10) {
                            StatTile(label: "Voltage",
                                     value: "\(Int(reading.gridVoltage.rounded()))",
                                     unit: "V",
                                     accent: .white)
                            StatTile(label: "Temp",
                                     value: "\(Int(reading.temperature.rounded()))",
                                     unit: "°C",
                                     accent: .white)
                        }
                    }

                    section("Raw") {
                        VStack(spacing: 0) {
                            InfoRow(label: "Timestamp", value: reading.timestampFormatted)
                            InfoRow(label: "Epoch seconds", value: "\(Int(reading.timestamp))")
                            InfoRow(label: "Cycle duration",
                                    value: String(format: "%.1f ms", reading.durationMs))
                        }
                        .padding(12)
                        .card()
                    }
                }
                .padding(20)
            }
            .background(Palette.backgroundTop)
            .navigationTitle("Reading detail")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button {
                        dismiss()
                    } label: {
                        Image(systemName: "xmark.circle.fill")
                            .foregroundStyle(.gray)
                    }
                    .accessibilityLabel("Close")
                }
            }
        }
        .preferredColorScheme(.dark)
    }

    private var header: some View {
        HStack(spacing: 12) {
            ZStack {
                RoundedRectangle(cornerRadius: 12)
                    .fill(Palette.inverterAmber.opacity(0.15))
                    .frame(width: 44, height: 44)
                Image(systemName: "waveform.path.ecg")
                    .font(.title3)
                    .foregroundStyle(Palette.inverterAmber)
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(reading.timestampFormatted)
                    .font(.title3.bold())
                    .foregroundStyle(.white)
                Text("Inverter reading")
                    .font(.caption)
                    .foregroundStyle(Palette.subtleText)
            }
            Spacer(minLength: 0)
        }
    }

    private func section<Content: View>(_ title: String, @ViewBuilder _ content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 10) {
            Text(title.uppercased())
                .font(.caption.weight(.semibold))
                .tracking(0.8)
                .foregroundStyle(Palette.subtleText)
            content()
        }
    }

    private var directionLabel: String {
        // battery_power is signed: positive = charging, negative = discharging.
        if reading.batteryPower > 5 { return "Charging" }
        if reading.batteryPower < -5 { return "Discharging" }
        return "Idle"
    }

    private func socColor(_ pct: Double) -> Color {
        if pct >= 50 { return Palette.battery }
        if pct >= 20 { return .orange }
        return .red
    }
}
